analyzers/src/SonarAnalyzer.VisualBasic/Rules/TooManyLabelsInSwitch.cs

```diff
@@ -18,31 +18,39 @@
 
-namespace SonarAnalyzer.Rules.VisualBasic
+namespace SonarAnalyzer.Rules.VisualBasic;
+
+[DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+public sealed class TooManyLabelsInSwitch : TooManyLabelsInSwitchBase<SyntaxKind, SelectStatementSyntax>
 {
-    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
-    public sealed class TooManyLabelsInSwitch : TooManyLabelsInSwitchBase<SyntaxKind, SelectStatementSyntax>
-    {
-        protected override DiagnosticDescriptor Rule { get; } =
-            DescriptorFactory.Create(DiagnosticId, MessageFormat,
-                isEnabledByDefault: false);
+    private static readonly ISet<SyntaxKind> IgnoredStatementsInSwitch = new HashSet<SyntaxKind> { SyntaxKind.ReturnStatement, SyntaxKind.ThrowStatement };
+
+    protected override DiagnosticDescriptor Rule { get; } =
+        DescriptorFactory.Create(DiagnosticId, string.Format(MessageFormat, "Select Case", "Case"),
+            isEnabledByDefault: false);
 
-        private const string MessageFormat = "Consider reworking this 'Select Case' to reduce the number of 'Case's" +
-            " from {1} to at most {0}.";
+    protected override SyntaxKind[] SyntaxKinds { get; } = [SyntaxKind.SelectStatement];
 
-        protected override SyntaxKind[] SyntaxKinds { get; } =
-            new[] { SyntaxKind.SelectStatement };
+    protected override GeneratedCodeRecognizer GeneratedCodeRecognizer =>
+        VisualBasicGeneratedCodeRecognizer.Instance;
 
-        protected override GeneratedCodeRecognizer GeneratedCodeRecognizer =>
-            VisualBasicGeneratedCodeRecognizer.Instance;
+    protected override SyntaxNode GetExpression(SelectStatementSyntax statement) =>
+        statement.Expression;
 
-        protected override SyntaxNode GetExpression(SelectStatementSyntax statement) =>
-            statement.Expression;
+    protected override int GetSectionsCount(SelectStatementSyntax statement) =>
+        ((SelectBlockSyntax)statement.Parent).CaseBlocks.Count;
 
-        protected override int GetSectionsCount(SelectStatementSyntax statement) =>
-            ((SelectBlockSyntax)statement.Parent).CaseBlocks.Count;
+    protected override bool AllSectionsAreOneLiner(SelectStatementSyntax statement) =>
+        ((SelectBlockSyntax)statement.Parent).CaseBlocks.All(HasOneLine);
+
+    protected override Location GetKeywordLocation(SelectStatementSyntax statement) =>
+        statement.SelectKeyword.GetLocation();
+
+    private static bool HasOneLine(CaseBlockSyntax switchSection)
+    {
+        var statements = switchSection.Statements.Where(x => !x.IsAnyKind(IgnoredStatementsInSwitch));
 
-        protected override Location GetKeywordLocation(SelectStatementSyntax statement) =>
-            statement.SelectKeyword.GetLocation();
+        return statements.Count() is 0
+               || (statements.Count() is 1 && statements.First().GetLocation().GetLineSpan().StartLinePosition.Line == statements.Last().GetLocation().GetLineSpan().EndLinePosition.Line);
```

Comment:
This logic is duplicated. I think it should be possible to extract at least parts of it (maybe generics help?).

Reply:
It is hard because there is no generic `ParametrizedDiagnosticAnalyzer` to provide the Language facade.
IMO, adding a generic `ParametrizedDiagnosticAnalyzer` is too much effort in the scope of this issue.